torch/csrc/api/src/optim/lbfgs.cpp

```diff
@@ -10,157 +10,513 @@
+      {at::indexing::Slice(offset, offset + numel)}).view_as(p), step_size);
     offset += numel;
   }
+  TORCH_INTERNAL_ASSERT(offset == _numel());
+}
+
+void LBFGS::_set_param(const std::vector<Tensor>& params_data) {
+  TORCH_INTERNAL_ASSERT(params_data.size() == _params.size());
+  for (size_t i = 0; i < _params.size(); i++) {
+    _params[i].copy_(params_data[i]);
+  }
+}
+
+std::vector<Tensor> LBFGS::_clone_param() {
+  std::vector<Tensor> result;
+  for (const auto& p : _params) {
+    result.emplace_back(p.clone(at::MemoryFormat::Contiguous));
+  }
+  return result;
 }
 
-torch::Tensor LBFGS::step(LossClosure closure) {
-  torch::Tensor orig_loss = closure();
-  torch::Tensor loss = orig_loss.clone(at::MemoryFormat::Contiguous);
-  int64_t current_evals = 1;
-  func_evals += 1;
+std::tuple<Tensor, Tensor> LBFGS::_directional_evaluate(
+  LossClosure closure, const std::vector<Tensor>& x, double t, const Tensor& d) {
+    _add_grad(t, d);
+    Tensor loss;
+    {
+      torch::AutoGradMode enable_grad(true);
+      loss = closure();
+    }
+    auto flat_grad = _gather_flat_grad();
+    _set_param(x);
+    return std::make_tuple(loss, flat_grad);
+}
 
-  Tensor flat_grad = gather_flat_grad();
-  Tensor abs_grad_sum = flat_grad.abs().sum();
+double _cubic_interpolate(
+  double x1, double f1, double g1, double x2, double f2, double g2,
+  c10::optional<std::tuple<double, double>> bounds = c10::nullopt) {
+  // ported from https://github.com/torch/optim/blob/master/polyinterp.lua
+  // Compute bounds of interpolation area
+  double xmin_bound, xmax_bound;
+  if (bounds != c10::nullopt) {
+    xmin_bound = std::get<0>(*bounds);
+    xmin_bound = std::get<1>(*bounds);
+  } else {
+    xmin_bound = (x1 <= x2) ? x1 : x2;
+    xmax_bound = (x1 <= x2) ? x2 : x1;
+  }
+  // Code for most common case: cubic interpolation of 2 points
+  //   w/ function and derivative values for both
+  // Solution in this case (where x2 is the farthest point):
+  //   d1 = g1 + g2 - 3*(f1-f2)/(x1-x2);
+  //   d2 = sqrt(d1^2 - g1*g2);
+  //   min_pos = x2 - (x2 - x1)*((g2 + d2 - d1)/(g2 - g1 + 2*d2));
+  //   t_new = min(max(min_pos,xmin_bound),xmax_bound);
 
-  if (abs_grad_sum.item<float>() <= options.tolerance_grad()) {
-    return loss;
+  auto d1 = g1 + g2 - 3 * (f1 - f2) / (x1 - x2);
+  auto d2_square = std::pow(d1, 2) - g1 * g2;
+  double d2;
+  if (d2_square >= 0) {
+    d2 = std::sqrt(d2_square);
+    double min_pos;
+    if(x1 <= x2) {
+      min_pos = x2 - (x2 - x1) * ((g2 + d2 - d1) / (g2 - g1 + 2 * d2));
+    } else {
+      min_pos = x1 - (x1 - x2) * ((g1 + d2 - d1) / (g1 - g2 + 2 * d2));
+    }
+    return std::min(std::max(min_pos, xmin_bound), xmax_bound);
+  } else {
+    return (xmin_bound + xmax_bound) / 2;
   }
+}
+
+template <typename Function>
+std::tuple<Tensor, Tensor, double, int64_t> _strong_wolfe(Function obj_func,
+      const std::vector<Tensor>& x, double t, Tensor& d, Tensor& f, Tensor& g, Tensor& gtd, double c1=1e-4,
```

Comment:
hmm why are `d` / `f` / `g` non-const references? (If we just want to be able to assign value to them in this function (and not affecting the original call site values), we can make them pass-by-value (e.g. `Tensor d, Tensor f, Tensor g`))